File: src/orchestrator/local_bridge/daemon.py
"""Polling daemon for the Local Agent Bridge."""

from __future__ import annotations

import argparse
import asyncio
import os
import socket
import uuid

import httpx

from orchestrator.local_bridge.executor import LocalExecutionError, LocalWorkspaceExecutor


class LocalBridgeDaemon:
    """Registers a local workspace and polls the orchestrator for local work."""

    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        bridge_id: str,
        workspace_root: str,
        poll_interval: float = 2.0,
        heartbeat_interval: float = 15.0,
        name: str = "lab-agentd",
        hostname: str | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._bridge_id = bridge_id
        self._workspace_root = os.path.abspath(workspace_root)
        self._poll_interval = poll_interval
        self._heartbeat_interval = heartbeat_interval
        self._name = name
        self._hostname = hostname or socket.gethostname()
        self._executor = LocalWorkspaceExecutor(self._workspace_root)

    @property
    def headers(self) -> dict[str, str]:
        return {"Authorization": f"Bearer {self._api_key}"}

    async def run(self) -> None:
        async with httpx.AsyncClient(timeout=30.0) as client:
            await self._register(client)
            last_heartbeat = 0.0
            while True:
                now = asyncio.get_running_loop().time()
                if now - last_heartbeat >= self._heartbeat_interval:
                    await self._heartbeat(client)
                    last_heartbeat = now

                claim = await self._claim_next(client)
                if claim is None:
                    await asyncio.sleep(self._poll_interval)
                    continue

                task_id = claim["task_id"]
                try:
                    result = await self._executor.execute(claim)
                except LocalExecutionError as exc:
                    result = {
                        "status": "error",
                        "summary": "Local execution rejected",
                        "error_message": str(exc),
                        "stderr": str(exc),
                    }
                await self._submit_result(client, task_id, result)

    async def _register(self, client: httpx.AsyncClient) -> None:
        response = await client.post(
            f"{self._base_url}/bridges/register",
            headers=self.headers,
            json={
                "bridge_id": self._bridge_id,
                "name": self._name,
                "hostname": self._hostname,
                "workspace_root": self._workspace_root,
                "capabilities": {
                    "tools": [
                        "read_file",
                        "list_files",
                        "write_file",
                        "apply_patch",
                        "run_command",
                        "git_status",
                        "git_diff",
                        "run_tests",
                    ]
                },
            },
        )
        response.raise_for_status()

    async def _heartbeat(self, client: httpx.AsyncClient) -> None:
        response = await client.post(
            f"{self._base_url}/bridges/{self._bridge_id}/heartbeat",
            headers=self.headers,
            json={"status": "active"},
        )
        response.raise_for_status()

    async def _claim_next(self, client: httpx.AsyncClient) -> dict | None:
        response = await client.post(
            f"{self._base_url}/bridges/{self._bridge_id}/claim-next",
            headers=self.headers,
        )
        response.raise_for_status()
        if not response.content or response.text == "null":
            return None
        return response.json()

    async def _submit_result(self, client: httpx.AsyncClient, task_id: str, result: dict) -> None:
        response = await client.post(
            f"{self._base_url}/bridges/{self._bridge_id}/tasks/{task_id}/result",
            headers=self.headers,
            json=result,
        )
        response.raise_for_status()


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Run the Local Agent Bridge daemon")
    parser.add_argument("--base-url", default=os.environ.get("LAB_AGENT_BASE_URL", "http://127.0.0.1:8100"))
    parser.add_argument("--api-key", default=os.environ.get("LAB_AGENT_API_KEY", ""))
    parser.add_argument("--bridge-id", default=os.environ.get("LAB_AGENT_BRIDGE_ID", str(uuid.uuid4())))
    parser.add_argument("--workspace-root", default=os.environ.get("LAB_AGENT_WORKSPACE_ROOT", os.getcwd()))
    parser.add_argument("--poll-interval", type=float, default=float(os.environ.get("LAB_AGENT_POLL_INTERVAL", "2")))
    parser.add_argument("--heartbeat-interval", type=float, default=float(os.environ.get("LAB_AGENT_HEARTBEAT_INTERVAL", "15")))
    parser.add_argument("--name", default=os.environ.get("LAB_AGENT_NAME", "lab-agentd"))
    return parser


def main() -> None:
    args = build_parser().parse_args()
    if not args.api_key:
        raise SystemExit("LAB_AGENT_API_KEY or --api-key is required")
    daemon = LocalBridgeDaemon(
        base_url=args.base_url,
        api_key=args.api_key,
        bridge_id=args.bridge_id,
        workspace_root=args.workspace_root,
        poll_interval=args.poll_interval,
        heartbeat_interval=args.heartbeat_interval,
        name=args.name,
    )
    asyncio.run(daemon.run())


if __name__ == "__main__":
    main()

File: internal/orchestratorgo/bridge/daemon.go
package bridge

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"

	"github.com/stratecode/lab/internal/orchestratorgo/domain"
)

type DaemonOptions struct {
	BaseURL           string
	APIKey            string
	BridgeID          string
	WorkspaceRoot     string
	Name              string
	Hostname          string
	PollInterval      time.Duration
	HeartbeatInterval time.Duration
}

type Daemon struct {
	client         *Client
	executor       *WorkspaceExecutor
	bridgeID       string
	workspaceRoot  string
	name           string
	hostname       string
	pollInterval   time.Duration
	heartbeatEvery time.Duration
	currentTaskID  *string
	currentStage   *string
	currentTool    *string
	currentSummary *string
}

func NewDaemon(opts DaemonOptions) (*Daemon, error) {
	workspaceRoot, err := filepath.Abs(firstNonEmptyString(strings.TrimSpace(opts.WorkspaceRoot), "."))
	if err != nil {
		return nil, err
	}
	executor, err := NewWorkspaceExecutor(workspaceRoot)
	if err != nil {
		return nil, err
	}
	bridgeID := stringsTrim(opts.BridgeID, uuid.NewString())
	name := stringsTrim(opts.Name, "lab-agentd")
	hostname := stringsTrim(opts.Hostname, defaultHostname())
	pollInterval := opts.PollInterval
	if pollInterval <= 0 {
		pollInterval = 2 * time.Second
	}
	heartbeatEvery := opts.HeartbeatInterval
	if heartbeatEvery <= 0 {
		heartbeatEvery = 15 * time.Second
	}
	return &Daemon{
		client:         NewClient(opts.BaseURL, opts.APIKey, 30*time.Second),
		executor:       executor,
		bridgeID:       bridgeID,
		workspaceRoot:  workspaceRoot,
		name:           name,
		hostname:       hostname,
		pollInterval:   pollInterval,
		heartbeatEvery: heartbeatEvery,
	}, nil
}

func (d *Daemon) Run(ctx context.Context) error {
	if _, err := d.client.Register(ctx, domain.LocalBridgeRegisterRequest{
		BridgeID:      d.bridgeID,
		Name:          d.name,
		Hostname:      d.hostname,
		WorkspaceRoot: d.workspaceRoot,
		Capabilities:  bridgeCapabilities(),
	}); err != nil {
		return err
	}
	lastHeartbeat := time.Time{}
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		if time.Since(lastHeartbeat) >= d.heartbeatEvery {
			if _, err := d.client.Heartbeat(ctx, d.bridgeID, "active", d.currentTaskID, intPtr(int(d.heartbeatEvery.Seconds())*3), d.currentStage, d.currentTool, d.currentSummary); err != nil {
				return err
			}
			lastHeartbeat = time.Now()
		}
		claim, err := d.client.ClaimNext(ctx, d.bridgeID)
		if err != nil {
			return err
		}
		if claim == nil {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(d.pollInterval):
				continue
			}
		}
		d.currentTaskID = &claim.TaskID
		d.currentStage = stringPtr("executing")
		d.currentTool = stringPtr(currentClaimTool(*claim))
		d.currentSummary = stringPtr(currentClaimSummary(*claim))
		type execOutcome struct {
			result  domain.LocalBridgeResultRequest
			execErr error
		}
		outcomeCh := make(chan execOutcome, 1)
		go func() {
			result, execErr := d.executor.Execute(ctx, *claim)
			outcomeCh <- execOutcome{result: result, execErr: execErr}
		}()
		var result domain.LocalBridgeResultRequest
		var execErr error
		executing := true
		for executing {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case outcome := <-outcomeCh:
				result = outcome.result
				execErr = outcome.execErr
				executing = false
			case <-time.After(d.heartbeatEvery):
				if _, err := d.client.Heartbeat(ctx, d.bridgeID, "active", d.currentTaskID, intPtr(int(d.heartbeatEvery.Seconds())*3), d.currentStage, d.currentTool, d.currentSummary); err != nil {
					return err
				}
				lastHeartbeat = time.Now()
			}
		}
		if execErr != nil {
			msg := execErr.Error()
			result = domain.LocalBridgeResultRequest{
				Status:       "error",
				Summary:      stringPtr("Local execution rejected"),
				Stderr:       &msg,
				ErrorMessage: &msg,
			}
		}
		if err := d.client.SubmitResult(ctx, d.bridgeID, claim.TaskID, result); err != nil {
			return fmt.Errorf("submit result for task %s: %w", claim.TaskID, err)
		}
		d.currentTaskID = nil
		d.currentStage = nil
		d.currentTool = nil
		d.currentSummary = nil
		log.Info().Str("bridge_id", d.bridgeID).Str("task_id", claim.TaskID).Str("status", result.Status).Msg("local bridge processed task")
	}
}

func currentClaimTool(claim domain.LocalBridgeTaskClaimResponse) string {
	toolRequest, ok := claim.Metadata["tool_request"].(map[string]any)
	if !ok {
		return ""
	}
	return strings.TrimSpace(asString(toolRequest["tool"]))
}

func currentClaimSummary(claim domain.LocalBridgeTaskClaimResponse) string {
	if summary := strings.TrimSpace(asString(claim.Metadata["coder_summary"])); summary != "" {
		return summary
	}
	if summary := strings.TrimSpace(asString(claim.Metadata["review_summary"])); summary != "" {
		return summary
	}
	return strings.TrimSpace(claim.Description)
}

func bridgeCapabilities() map[string]any {
	return map[string]any{
		"tools": []string{
			"read_file",
			"filesystem.read",
			"list_files",
			"filesystem.list",
			"write_file",
			"filesystem.write",
			"research_project",
			"scaffold_project",
			"review_project",
			"review_workspace",
			"apply_patch",
			"run_command",
			"git_status",
			"git_diff",
			"run_tests",
		},
	}
}

func defaultHostname() string {
	name, err := os.Hostname()
	if err != nil || stringsTrim(name, "") == "" {
		return "localhost"
	}
	return name
}

func stringsTrim(value, fallback string) string {
	if strings.TrimSpace(value) == "" {
		return fallback
	}
	return strings.TrimSpace(value)
}
